circuit/src/gadgets/sum.rs

```diff
@@ -0,0 +1,226 @@
+				let mut sum = region.assign_advice_from_constant(
+					|| "initial_sum",
+					config.sum,
+					0,
+					F::zero(),
+				)?;
+
+				// This circuit can take many inputs. It will loop and accumulate all of the
+				// items to the sum value.
+				for i in 0..S {
+					config.selector.enable(&mut region, i)?;
+					let item =
+						self.items[i].copy_advice(|| "item", &mut region, config.items, i)?;
+					let val = sum.value().cloned() + item.value();
+					sum = region.assign_advice(|| "sum", config.sum, i + 1, || val)?;
+				}
+
+				Ok(sum)
+			},
+		)
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use std::usize;
+
+	use super::*;
+	use crate::utils::{generate_params, prove_and_verify};
+	use halo2wrong::{
+		curves::bn256::{Bn256, Fr},
+		halo2::{
+			arithmetic::Field,
+			circuit::{SimpleFloorPlanner, Value},
+			dev::MockProver,
+			plonk::{Circuit, Instance},
+		},
+	};
+
+	#[derive(Clone)]
+	struct TestConfig {
+		sum: SumConfig,
+		temp: Column<Advice>,
+		pub_ins: Column<Instance>,
+	}
+
+	#[derive(Clone)]
+	struct TestCircuit<F: FieldExt, const S: usize> {
+		items: [F; S],
+	}
+
+	impl<F: FieldExt, const S: usize> TestCircuit<F, S> {
+		fn new(items: [F; S]) -> Self {
+			Self { items }
+		}
+	}
+
+	impl<F: FieldExt, const S: usize> Circuit<F> for TestCircuit<F, S> {
+		type Config = TestConfig;
+		type FloorPlanner = SimpleFloorPlanner;
+
+		fn without_witnesses(&self) -> Self {
+			self.clone()
+		}
+
+		fn configure(meta: &mut ConstraintSystem<F>) -> TestConfig {
+			let sum = SumChip::<_, S>::configure(meta);
+			let temp = meta.advice_column();
+			let pub_ins = meta.instance_column();
+
+			meta.enable_equality(temp);
+			meta.enable_equality(pub_ins);
+
+			TestConfig { sum, temp, pub_ins }
+		}
+
+		fn synthesize(
+			&self,
+			config: TestConfig,
+			mut layouter: impl Layouter<F>,
+		) -> Result<(), Error> {
+			let arr = layouter.assign_region(
+				|| "temp",
+				|mut region: Region<'_, F>| {
+					let mut arr: [Option<AssignedCell<F, F>>; S] = [(); S].map(|_| None);
+					for i in 0..S {
+						arr[i] = Some(region.assign_advice(
+							|| "temp",
+							config.temp,
+							i,
+							|| Value::known(self.items[i]),
+						)?);
+					}
+					Ok(arr.map(|a| a.unwrap()))
+				},
+			)?;
+			let sum_chip = SumChip::new(arr);
+			let acc = sum_chip.synthesize(config.sum, layouter.namespace(|| "sum"))?;
+
+			layouter.constrain_instance(acc.cell(), config.pub_ins, 0)?;
+			Ok(())
+		}
+	}
+
+	#[test]
+	fn test_sum() {
+		// Testing three 2's representation on the field as an input.
+		let test_chip = TestCircuit::new([Fr::from(2); 3]);
+
+		let k = 4;
+		let pub_ins = vec![Fr::from(6)];
+		let prover = MockProver::run(k, &test_chip, vec![pub_ins]).unwrap();
+		assert_eq!(prover.verify(), Ok(()));
+	}
+
+	#[test]
+	fn test_sum_random_element() {
+		// Testing big array with a random element from the field.
+		let rng = &mut rand::thread_rng();
+		let array = [<Fr as Field>::random(rng); 20];
```

Comment:
This will create 20 of the same field values. We want every value to be different.
And let's try bigger arrays, so perhaps try:
```rust
let array = [(); 512].map(|_| <Fr as Field>::random(rng));
```

Reply:
But in that case how I can calculate the public input? Should I use different calculators to get the public input and test the circuit?

Reply:
Yes, loop through the array and calculate the sum, and the result as a public input